File: raw-compiler-rql2-truffle/src/main/scala/raw/compiler/rql2/truffle/builtin/TruffleAwsPackage.scala
/*
 * Copyright 2023 RAW Labs S.A.
 *
 * Use of this software is governed by the Business Source License
 * included in the file licenses/BSL.txt.
 *
 * As of the Change Date specified in that file, in accordance with
 * the Business Source License, use of this software will be governed
 * by the Apache License, Version 2.0, included in the file
 * licenses/APL.txt.
 */

package raw.compiler.rql2.truffle.builtin

import raw.compiler.base.source.Type
import raw.compiler.rql2.builtin.AwsV4SignedRequest
import raw.compiler.rql2.source.{Rql2ListType, Rql2RecordType}
import raw.compiler.rql2.truffle.{TruffleArg, TruffleEntryExtension}
import raw.runtime.truffle.{ExpressionNode, RawLanguage}
import raw.runtime.truffle.ast.expressions.binary.PlusNode
import raw.runtime.truffle.ast.expressions.builtin.aws_package.AwsV4SignedRequestNodeGen
import raw.runtime.truffle.ast.expressions.iterable.list.ListBuildNode
import raw.runtime.truffle.ast.expressions.literals.StringNode

class AwsV4SignedRequestEntry extends AwsV4SignedRequest with TruffleEntryExtension {

  override def toTruffle(t: Type, args: Seq[TruffleArg], rawLanguage: RawLanguage): ExpressionNode = {

    val key = args.head.e
    val secretKey = args(1).e
    val service = args(2).e

    val maybeRegion = args.collectFirst { case TruffleArg(e, _, Some(idn)) if idn == "region" => e }
    val maybeSessionToken = args.collectFirst { case TruffleArg(e, _, Some(idn)) if idn == "sessionToken" => e }
    val method =
      args.collectFirst { case TruffleArg(e, _, Some(idn)) if idn == "method" => e }.getOrElse(new StringNode("GET"))

    val host = args
      .collectFirst { case TruffleArg(e, _, Some(idn)) if idn == "host" => e }
      .getOrElse(
        if (maybeRegion.isDefined) new PlusNode(
          new PlusNode(new PlusNode(service, new StringNode(".")), maybeRegion.get),
          new StringNode(".amazonaws.com")
        )
        else new PlusNode(service, new StringNode(".amazonaws.com"))
      )

    val path =
      args.collectFirst { case TruffleArg(e, _, Some(idn)) if idn == "path" => e }.getOrElse(new StringNode("/"))

    val body =
      args.collectFirst { case TruffleArg(e, _, Some(idn)) if idn == "bodyString" => e }.getOrElse(new StringNode(""))

    val urlParams = args
      .collectFirst { case TruffleArg(e, _, Some(idn)) if idn == "args" => e }
      .getOrElse(new ListBuildNode(Rql2ListType(Rql2RecordType(Vector.empty)), Array()))

    val headers = args
      .collectFirst { case TruffleArg(e, _, Some(idn)) if idn == "headers" => e }
      .getOrElse(new ListBuildNode(Rql2ListType(Rql2RecordType(Vector())), Array()))

    val sessionToken = maybeSessionToken
      .getOrElse(new StringNode(""))

    val region = maybeRegion.getOrElse(new StringNode("us-east-1"))

    AwsV4SignedRequestNodeGen.create(
      key,
      secretKey,
      service,
      region,
      sessionToken,
      path,
      method,
      host,
      body,
      urlParams,
      headers
    )
  }

}

File: extensions/src/main/java/raw/compiler/snapi/truffle/builtin/TruffleAwsPackage.java
package raw.compiler.snapi.truffle.builtin;

import java.util.List;
import java.util.Optional;
import raw.compiler.base.source.Type;
import raw.compiler.rql2.builtin.AwsV4SignedRequest;
import raw.compiler.rql2.source.Rql2ListType;
import raw.compiler.rql2.source.Rql2RecordType;
import raw.compiler.snapi.truffle.TruffleArg;
import raw.compiler.snapi.truffle.TruffleEntryExtension;
import raw.runtime.truffle.ExpressionNode;
import raw.runtime.truffle.RawLanguage;
import raw.runtime.truffle.ast.expressions.binary.PlusNode;
import raw.runtime.truffle.ast.expressions.builtin.aws_package.AwsV4SignedRequestNodeGen;
import raw.runtime.truffle.ast.expressions.iterable.list.ListBuildNode;
import raw.runtime.truffle.ast.expressions.literals.StringNode;
import scala.collection.immutable.HashSet;
import scala.collection.immutable.Vector;

public class TruffleAwsPackage extends AwsV4SignedRequest implements TruffleEntryExtension {
  @Override
  public ExpressionNode toTruffle(Type type, List<TruffleArg> args, RawLanguage rawLanguage) {
    ExpressionNode key = args.get(0).getExpNode();
    ExpressionNode secretKey = args.get(1).getExpNode();
    ExpressionNode service = args.get(2).getExpNode();

    Optional<ExpressionNode> maybeRegion =
        args.stream()
            .filter(
                (TruffleArg a) -> a.getIdentifier() != null && a.getIdentifier().equals("region"))
            .map(TruffleArg::getExpNode)
            .findFirst();

    Optional<ExpressionNode> maybeSessionToken =
        args.stream()
            .filter(
                (TruffleArg a) ->
                    a.getIdentifier() != null && a.getIdentifier().equals("sessionToken"))
            .map(TruffleArg::getExpNode)
            .findFirst();

    Optional<ExpressionNode> maybeMethod =
        args.stream()
            .filter(
                (TruffleArg a) -> a.getIdentifier() != null && a.getIdentifier().equals("method"))
            .map(TruffleArg::getExpNode)
            .findFirst();

    ExpressionNode method = maybeMethod.orElseGet(() -> new StringNode("GET"));

    Optional<ExpressionNode> maybeHost =
        args.stream()
            .filter((TruffleArg a) -> a.getIdentifier() != null && a.getIdentifier().equals("host"))
            .map(TruffleArg::getExpNode)
            .findFirst();

    ExpressionNode host =
        maybeHost.orElseGet(
            () ->
                maybeRegion
                    .map(
                        expressionNode ->
                            new PlusNode(
                                new PlusNode(
                                    new PlusNode(service, new StringNode(".")), expressionNode),
                                new StringNode(".amazonaws.com")))
                    .orElseGet(() -> new PlusNode(service, new StringNode(".amazonaws.com"))));

    ExpressionNode path =
        args.stream()
            .filter((TruffleArg a) -> a.getIdentifier() != null && a.getIdentifier().equals("path"))
            .map(TruffleArg::getExpNode)
            .findFirst()
            .orElseGet(() -> new StringNode("/"));

    ExpressionNode body =
        args.stream()
            .filter(
                (TruffleArg a) ->
                    a.getIdentifier() != null && a.getIdentifier().equals("bodyString"))
            .map(TruffleArg::getExpNode)
            .findFirst()
            .orElseGet(() -> new StringNode(""));

    ExpressionNode urlParams =
        args.stream()
            .filter((TruffleArg a) -> a.getIdentifier() != null && a.getIdentifier().equals("args"))
            .map(TruffleArg::getExpNode)
            .findFirst()
            .orElseGet(
                () ->
                    new ListBuildNode(
                        Rql2ListType.apply(
                            Rql2RecordType.apply(Vector.empty(), new HashSet<>()), new HashSet<>()),
                        new ExpressionNode[] {}));

    ExpressionNode headers =
        args.stream()
            .filter(
                (TruffleArg a) -> a.getIdentifier() != null && a.getIdentifier().equals("headers"))
            .map(TruffleArg::getExpNode)
            .findFirst()
            .orElseGet(
                () ->
                    new ListBuildNode(
                        Rql2ListType.apply(
                            Rql2RecordType.apply(Vector.empty(), new HashSet<>()), new HashSet<>()),
                        new ExpressionNode[] {}));

    ExpressionNode sessionToken = maybeSessionToken.orElseGet(() -> new StringNode(""));

    ExpressionNode region = maybeRegion.orElseGet(() -> new StringNode("us-east-1"));

    return AwsV4SignedRequestNodeGen.create(
        key,
        secretKey,
        service,
        region,
        sessionToken,
        path,
        method,
        host,
        body,
        urlParams,
        headers);
  }
}
